msprime/branch_stats.py

```diff
@@ -0,0 +1,229 @@
+
+def branch_length_Y(ts,x,y,z):
+    S = 0
+    for tr in ts.trees():
+        xy_mrca = tr.mrca(x,y)
+        xz_mrca = tr.mrca(x,z)
+        yz_mrca = tr.mrca(y,z)
+        if xy_mrca == xz_mrca:
+            #   /\
+            #  / /\
+            # x y  z
+            S += path_length(tr,x,yz_mrca)*tr.length
+        elif xy_mrca == yz_mrca:
+            #   /\
+            #  / /\
+            # y x  z
+            S += path_length(tr,x,xz_mrca)*tr.length
+        elif xz_mrca == yz_mrca:
+            #   /\
+            #  / /\
+            # z x  y
+            S += path_length(tr,x,xy_mrca)*tr.length
+    return S/ts.sequence_length
+
+def branch_length_f4(ts,A,B,C,D):
+    for U in A,B,C,D:
+        if max([U.count(x) for x in set(U)])>1:
+            raise ValueError("A,B,C, and D cannot contain repeated elements.")
+    S = 0
+    for tr in ts.trees():
+        SS = 0
+        for a in A:
+            for b in B:
+                for c in C:
+                    for d in D:
+                        SS += path_length(tr,tr.mrca(a,c),tr.mrca(b,d))
+                        SS -= path_length(tr,tr.mrca(a,d),tr.mrca(b,c))
+        S += SS*tr.length
+    return S/(ts.sequence_length*len(A)*len(B)*len(C)*len(D))
+
+def branch_stats_node_iter(ts,leaf_sets,weight_fun,method='length'):
+    '''
+    Here leaf_sets is a list of lists of leaves, and weight_fun is a function
+    whose argument is a list of integers of the same length as leaf_sets
+    that returns a number.  Each branch in a tree is weighted by weight_fun(x),
+    where x[i] is the number of leaves in leaf_sets[i] below that
+    branch.  This finds the sum of all counted branches for each tree,
+    and averages this across the tree sequence ts, weighted by genomic length.
+
+    If method='mutations' instead, then branch lengths will be measured in
+    numbers of mutations instead of time.
+
+    This version is inefficient as it iterates over all nodes in each tree.
+    '''
+    out = branch_stats_vector_node_iter(ts,leaf_sets,lambda x: [weight_fun(x)],method)
+    if len(out)>1:
+        raise ValueError("Expecting output of length 1.")
+    return out[0]
+
+def branch_stats(ts,leaf_sets,weight_fun,method='length'):
+    '''
+    Here leaf_sets is a list of lists of leaves, and weight_fun is a function
+    whose argument is a list of integers of the same length as leaf_sets
+    that returns a boolean.  A branch in a tree is weighted by weight_fun(x),
+    where x[i] is the number of leaves in leaf_sets[i] below that
+    branch.  This finds the sum of all counted branches for each tree,
+    and averages this across the tree sequence ts, weighted by genomic length.
+
+    Doesn't do method='mutations'.
+    '''
+    out = branch_stats_vector(ts,leaf_sets,lambda x: [weight_fun(x)],method)
+    if len(out)>1:
+        raise ValueError("Expecting output of length 1.")
+    return out[0]
+
+def branch_stats_vector_node_iter(ts,leaf_sets,weight_fun,method='length'):
```

Comment:
Ultimately, ``branch_stats`` (or something similar) will be a method of the tree sequence. I think we should offer the general interface as something like ``branch_stats(self, leaf_sets, weight_fun)``, and use the most efficient algorithm we have implemented in Python. So, let's move the ``branch_stats_vector_node_iter`` function into the TreeSequence, and move the less efficient implementations into the ``test_branch_stats.py``.

I think we should drop the 'method=whatever' argument for now. Once we have the ``num_branch_mutations`` method (#115), this will be easy to add.